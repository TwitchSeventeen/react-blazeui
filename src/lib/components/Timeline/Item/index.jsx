import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

function Item(props) {
  const {
    children, as: Component, className, last, type, loading, left, dataName, ...other
  } = props;
  const classes = classNames(
    'c-timeline-item',
    { 'c-timeline-item--last': last },
    { [`c-timeline-item--${type}`]: type },
    { 'c-timeline-item--loading': loading },
    { 'c-timeline-item--left': left },
    className,
  );

  return (
    <Component
      {...other}
      className={classes}
      data-name={dataName}
    >
      <div className="c-timeline-item__body">
        { children }
      </div>
    </Component>
  );
}

Item.propTypes = {
  /**
   * This element can have children
   */
  children: PropTypes.node,
  /**
   * Additional classes
   */
  className: PropTypes.string,
  /**
   * Identifies the last item in a timeline
   */
  last: PropTypes.bool,
  /**
   * The type of item
   */
  type: PropTypes.oneOf([
    'brand',
    'info',
    'warning',
    'success',
    'error',
  ]),
  /**
   * Defines an item that hasn't happened yet
   */
  loading: PropTypes.bool,
  /**
   * Displays an item on the left side of the Timeline
   */
  left: PropTypes.bool,
  /**
   * Optional data-name attribute
   * (use for testing/selection)
   */
  dataName: PropTypes.string,
  /**
   * Element type to render as
   */
  as: PropTypes.elementType,
};

Item.defaultProps = {
  children: null,
  className: null,
  last: false,
  type: null,
  loading: false,
  left: false,
  dataName: 'timeline-item',
  as: 'li',
};

export default Item;
