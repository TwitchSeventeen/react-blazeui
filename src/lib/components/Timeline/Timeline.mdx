---
name: Timeline
route: /components/timeline
menu: Components
---

import { Playground, PropsTable } from 'docz';
import Timeline from './index';
import Quiet from '../Quiet';

# Timeline
Timelines are good to illustrate a connected set of events that have happened or are planned to happen.

## Standard
<Playground>
    <Timeline>
        <Timeline.Item>
            Nothing
        </Timeline.Item>
        <Timeline.Item>
            Big Bang
        </Timeline.Item>
        <Timeline.Item>
            Galaxies go everywhere
        </Timeline.Item>
        <Timeline.Item>
            Solar systems form
        </Timeline.Item>
        <Timeline.Item>
            Earth appears
        </Timeline.Item>
        <Timeline.Item>
            Human infestation occurs!
        </Timeline.Item>
        <Timeline.Item last>
            All this has happened before, and will happen again
        </Timeline.Item>
    </Timeline>
</Playground>

## Colours
<Playground>
    <Timeline>
        <Timeline.Item>
            Nothing
        </Timeline.Item>
        <Timeline.Item type="brand">
            Big Bang
        </Timeline.Item>
        <Timeline.Item type="info">
            Galaxies go everywhere
        </Timeline.Item>
        <Timeline.Item type="warning">
            Solar systems form
        </Timeline.Item>
        <Timeline.Item type="success">
            Earth appears
        </Timeline.Item>
        <Timeline.Item type="error">
            Human infestation occurs!
        </Timeline.Item>
        <Timeline.Item last>
            All this has happened before, and will happen again
        </Timeline.Item>
    </Timeline>
</Playground>

## Loading last item

Use the **loading** prop on both the _Timeline_ and _Timeline.Item_ components to indicate the last item hasn't happened yet.
<Playground>
    <Timeline loading>
        <Timeline.Item>
            Nothing
        </Timeline.Item>
        <Timeline.Item>
            Big Bang
        </Timeline.Item>
        <Timeline.Item>
            Galaxies go everywhere
        </Timeline.Item>
        <Timeline.Item>
            Solar systems form
        </Timeline.Item>
        <Timeline.Item>
            Earth appears
        </Timeline.Item>
        <Timeline.Item>
            Human infestation occurs!
        </Timeline.Item>
        <Timeline.Item last>
            All this has happened before, and will happen again
        </Timeline.Item>
        <Timeline.Item loading>
            ...Evolution
        </Timeline.Item>
    </Timeline>
</Playground>

## Alternating items

Use the **left** prop on a _Timeline.Item_ to put it on the left side of the Timeline.

Make sure to put the **alternate** prop on the _Timeline_ as well.

<Playground>
    <Timeline loading alternate>
        <Timeline.Item>
            Nothing
        </Timeline.Item>
        <Timeline.Item type="brand">
            Big Bang
        </Timeline.Item>
        <Timeline.Item type="info" left>
            Galaxies go everywhere
        </Timeline.Item>
        <Timeline.Item type="warning">
            Solar systems form
        </Timeline.Item>
        <Timeline.Item type="success" left>
            Earth appears
        </Timeline.Item>
        <Timeline.Item type="error">
            Human infestation occurs!
        </Timeline.Item>
        <Timeline.Item last left>
            All this has happened before, and will happen again
        </Timeline.Item>
        <Timeline.Item loading>
            ...Evolution
        </Timeline.Item>
    </Timeline>
</Playground>

## Properties
### Timeline
<PropsTable of={Timeline} />

### Timeline.Item
<PropsTable of={Timeline.Item} />