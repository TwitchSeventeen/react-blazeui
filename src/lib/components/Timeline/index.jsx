import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import Item from './Item';

function Timeline(props) {
  const {
    children, className, loading, alternate, dataName, ...other
  } = props;
  const classes = classNames(
    'o-timeline',
    { 'o-timeline--loading': loading },
    { 'o-timeline--alternate': alternate },
    className,
  );

  return (
    <ul
      {...other}
      className={classes}
      data-name={dataName}
    >
      { children }
    </ul>
  );
}

Timeline.propTypes = {
  /**
   * This element can have children
   */
  children: PropTypes.node,
  /**
   * Additional classes
   */
  className: PropTypes.string,
  /**
   * This Timeline contains Item's that haven't happened yet
   */
  loading: PropTypes.bool,
  /**
   * The items on this Timeline can alternate between left and right side
   */
  alternate: PropTypes.bool,
  /**
   * Optional data-name attribute
   * (use for testing/selection)
   */
  dataName: PropTypes.string,
};

Timeline.defaultProps = {
  children: null,
  className: null,
  loading: false,
  alternate: false,
  dataName: 'timeline',
};

Timeline.Item = Item;

export default Timeline;
