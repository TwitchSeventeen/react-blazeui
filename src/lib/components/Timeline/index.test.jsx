import React from 'react';
import { render, cleanup } from '@testing-library/react';
import Timeline from './index';

describe('Timeline', () => {
  afterEach(cleanup);

  it('should be defined', () => {
    expect(Timeline).toBeDefined();
  });

  it('renders correctly', () => {
    const { asFragment } = render(
      <Timeline>
        <Timeline.Item>
          lorem ipsum dolor
        </Timeline.Item>
      </Timeline>,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders correctly with a loading item', () => {
    const { asFragment } = render(
      <Timeline loading>
        <Timeline.Item>
          lorem ipsum dolor
        </Timeline.Item>
        <Timeline.Item loading>
          lorem ipsum dolor
        </Timeline.Item>
      </Timeline>,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders correctly with left Items', () => {
    const { asFragment } = render(
      <Timeline alternate>
        <Timeline.Item>
          lorem ipsum dolor
        </Timeline.Item>
        <Timeline.Item left>
          lorem ipsum dolor
        </Timeline.Item>
      </Timeline>,
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
