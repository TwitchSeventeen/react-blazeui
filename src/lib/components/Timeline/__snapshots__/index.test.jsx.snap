// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timeline renders correctly 1`] = `
<DocumentFragment>
  <ul
    class="o-timeline"
    data-name="timeline"
  >
    <li
      class="c-timeline-item"
      data-name="timeline-item"
    >
      <div
        class="c-timeline-item__body"
      >
        lorem ipsum dolor
      </div>
    </li>
  </ul>
</DocumentFragment>
`;

exports[`Timeline renders correctly with a loading item 1`] = `
<DocumentFragment>
  <ul
    class="o-timeline o-timeline--loading"
    data-name="timeline"
  >
    <li
      class="c-timeline-item"
      data-name="timeline-item"
    >
      <div
        class="c-timeline-item__body"
      >
        lorem ipsum dolor
      </div>
    </li>
    <li
      class="c-timeline-item c-timeline-item--loading"
      data-name="timeline-item"
    >
      <div
        class="c-timeline-item__body"
      >
        lorem ipsum dolor
      </div>
    </li>
  </ul>
</DocumentFragment>
`;

exports[`Timeline renders correctly with left Items 1`] = `
<DocumentFragment>
  <ul
    class="o-timeline o-timeline--alternate"
    data-name="timeline"
  >
    <li
      class="c-timeline-item"
      data-name="timeline-item"
    >
      <div
        class="c-timeline-item__body"
      >
        lorem ipsum dolor
      </div>
    </li>
    <li
      class="c-timeline-item c-timeline-item--left"
      data-name="timeline-item"
    >
      <div
        class="c-timeline-item__body"
      >
        lorem ipsum dolor
      </div>
    </li>
  </ul>
</DocumentFragment>
`;
